pkg/metricshandler/metrics_handler.go

```diff
@@ -0,0 +1,243 @@
+/*
+Copyright 2019 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metricshandler
+
+import (
+	"compress/gzip"
+	"context"
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/pkg/errors"
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+	"k8s.io/kube-state-metrics/internal/store"
+	metricsstore "k8s.io/kube-state-metrics/pkg/metrics_store"
+	"k8s.io/kube-state-metrics/pkg/options"
+)
+
+type MetricsHandler struct {
+	opts               *options.Options
+	kubeClient         kubernetes.Interface
+	storeBuilder       *store.Builder
+	enableGZIPEncoding bool
+
+	cancel func()
+
+	// mtx protects stores, curShard, and curTotalShards
+	mtx            *sync.RWMutex
+	stores         []*metricsstore.MetricsStore
+	curShard       int32
+	curTotalShards int
+}
+
+func New(opts *options.Options, kubeClient kubernetes.Interface, storeBuilder *store.Builder, enableGZIPEncoding bool) *MetricsHandler {
+	return &MetricsHandler{
+		opts:               opts,
+		kubeClient:         kubeClient,
+		storeBuilder:       storeBuilder,
+		enableGZIPEncoding: enableGZIPEncoding,
+		mtx:                &sync.RWMutex{},
+	}
+}
+
+func (m *MetricsHandler) ConfigureSharding(ctx context.Context, shard int32, totalShards int) {
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+
+	if m.cancel != nil {
+		m.cancel()
+	}
+	klog.Infof("configuring sharding of this instance to be shard index %d (zero-indexed) out of %d total shards", shard, totalShards)
+	ctx, m.cancel = context.WithCancel(ctx)
+	m.storeBuilder.WithSharding(shard, totalShards)
+	m.storeBuilder.WithContext(ctx)
+	m.stores = m.storeBuilder.Build()
+	m.curShard = shard
+	m.curTotalShards = totalShards
+}
+
+func (m *MetricsHandler) Run(ctx context.Context) error {
+	autoSharding := len(m.opts.Pod) > 0 && len(m.opts.Namespace) > 0
+
+	klog.Infof("autosharding enabled: %t; pod=%v; pod_namespace=%v", autoSharding, m.opts.Pod, m.opts.Namespace)
```

Comment:
Hmm not sure I like that this log is always logging. Also is it true that autosharding is enabled, how do we know that? From the options warning, it says you need to enable it? Can you elaborate, thanks! 

Nits: Also can we do the following:
- set this log level to debug
- default values for those fields are empty strings, meaning this msg will not make much sense.

Reply:
It's only logged once at startup and is very valuable information, I'd prefer to keep it at info level. Otherwise I split the log line into two (one enabled, one disabled) to make it less confusing.